src/t_hash.c

```diff
@@ -209,25 +210,222 @@ static void hashDictWithExpireOnRelease(dict *d) {
+    lpFree(lpt->lp);
+    zfree(lpt);
+}
+
+/* Returns number of expired fields. */
+static uint64_t listpackTTLExpireDryRun(const robj *o) {
+    serverAssert(o->encoding == OBJ_ENCODING_LISTPACK_TTL);
+
+    uint64_t expired = 0;
+    unsigned char *fptr, *s;
+    listpackTTL *lpt = o->ptr;
+
+    fptr = lpFirst(lpt->lp);
+    while (fptr != NULL) {
+        long long val;
+
+        fptr = lpNext(lpt->lp, fptr);
+        serverAssert(fptr);
+        fptr = lpNext(lpt->lp, fptr);
+        serverAssert(fptr);
+
+        s = lpGetValue(fptr, NULL, &val);
+        serverAssert(!s);
+
+        if (hashTypeListpackIsExpired(val))
+            expired++;
+
+        fptr = lpNext(lpt->lp, fptr);
+    }
+
+    return expired;
+}
+
+/* Returns the expiration time of the item with the nearest expiration. */
+static uint64_t listpackTTLGetMinExpire(robj *o) {
+    serverAssert(o->encoding == OBJ_ENCODING_LISTPACK_TTL);
+
+    uint64_t minExpire = EB_EXPIRE_TIME_INVALID;
+    unsigned char *fptr, *s;
+    listpackTTL *lpt = o->ptr;
+
+    fptr = lpFirst(lpt->lp);
+    while (fptr != NULL) {
+        long long val;
+
+        fptr = lpNext(lpt->lp, fptr);
+        serverAssert(fptr);
+        fptr = lpNext(lpt->lp, fptr);
+        serverAssert(fptr);
+
+        s = lpGetValue(fptr, NULL, &val);
+        serverAssert(!s);
+
+        if (val != HASH_LP_NO_TTL && (uint64_t) val < minExpire)
+            minExpire = val;
+
+        fptr = lpNext(lpt->lp, fptr);
+    }
+
+    return minExpire;
+}
+
+/* Walk over fields and delete the expired ones. */
+static void listpackTTLExpire(robj *o, ExpireInfo *info) {
+    serverAssert(o->encoding == OBJ_ENCODING_LISTPACK_TTL);
+
+    long long min = LLONG_MAX;
+    unsigned char *ptr, *field, *s;
+    listpackTTL *lpt = o->ptr;
+
+    ptr = lpFirst(lpt->lp);
+    /* TODO: Optimize this loop. 'maxToExpire' does not make much sense for listpack. */
+    while (ptr != NULL && (info->itemsExpired < info->maxToExpire)) {
+        long long val;
+
+        field = ptr;
+        ptr = lpNext(lpt->lp, ptr);
+        serverAssert(ptr);
+        ptr = lpNext(lpt->lp, ptr);
+        serverAssert(ptr);
+
+        s = lpGetValue(ptr, NULL, &val);
+        serverAssert(!s);
+
+        if (val != HASH_LP_NO_TTL && (uint64_t) val < info->now) {
+            server.stat_expired_hash_fields++;
+            lpt->lp = lpDeleteRangeWithEntry(lpt->lp, &field, 3);
+            ptr = field;
+            info->itemsExpired++;
+        } else {
+            if (val != HASH_LP_NO_TTL && val < min)
+                min = val;
+
+            ptr = lpNext(lpt->lp, ptr);
+        }
+    }
+
+    /* If we visited all the fields, ptr should be NULL and `min` should contain
+     * the minimum expiration time. */
+    if (!ptr) {
+        info->nextExpireTime = (min == LLONG_MAX) ? 0 : min;
+    } else {
+        /* TODO: Slower path, check performance */
+        info->nextExpireTime = hashTypeGetNextTimeToExpire(o);
```

Comment:
It is expensive. You can respect `info->maxToExpire` and provide `nextExpireTime` in the same `while` loop. Something like:
```
    while (ptr != NULL) {
        ....
        if ((val != HASH_LP_NO_TTL) && ((uint64_t) val < info->now) && (info->itemsExpired < info->maxToExpire)) {
            server.stat_expired_hash_fields++;
            lpt->lp = lpDeleteRangeWithEntry(lpt->lp, &field, 3);
            ptr = field;
            info->itemsExpired++;
        } else {
            if (val != HASH_LP_NO_TTL && val < min)
                min = val;
            ptr = lpNext(lpt->lp, ptr);
        }
    }
    info->nextExpireTime = (min == LLONG_MAX) ? 0 : min;
```